kernel/src/drivers/ps2/io.rs

```diff
@@ -1,132 +1,139 @@
+        WriteInputPort2 = 0xD4,
+    }
 
-pub const WAIT_TIMEOUT: u16 = 1000;
+    /// Represents a PS2 controller command with a return value
+    #[allow(dead_code)] // Dead variants for completeness
+    #[derive(Copy, Clone)]
+    #[repr(u8)]
+    pub enum ControllerReturnCommand {
+        ReadConfig = 0x20,
+        TestController = 0xAA,
+        TestPort1 = 0xAB,
+        TestPort2 = 0xA9,
+        IdentifyDevice = 0xF2,
+    }
 
-const OUTPUT_STATUS_BIT: u8 = 1 << 0;
-const INPUT_STATUS_BIT: u8 = 1 << 1;
+    /// Represents a PS2 controller command with a data value
+    #[allow(dead_code)] // Dead variants for completeness
+    #[derive(Copy, Clone)]
+    #[repr(u8)]
+    pub enum ControllerDataCommand {
+        WriteConfig = 0x60,
+    }
 
-/// Sends a controller command without a return
-pub fn command(cmd: ControllerCommand) {
-    command_raw(cmd as u8);
+    /// Represents a PS2 device command opcode
+    #[allow(dead_code)] // Dead variants for completeness
+    #[derive(Copy, Clone)]
+    #[repr(u8)]
+    pub enum DeviceCommand {
+        EnableScanning = 0xF4,
+        DisableScanning = 0xF5,
+        SetDefaults = 0xF6,
+        SetScancode = 0xF0,
+        Reset = 0xFF,
+    }
 }
 
-/// Sends a controller command with data and without a return
-pub fn command_data(cmd: ControllerCommand, data: u8) {
-    command_raw(cmd as u8);
-    write(&DATA_PORT, data);
-}
+use io::IOPort;
 
-/// Sends a controller command with a return
-pub fn command_ret(cmd: ControllerReturnCommand) -> Option<u8> {
-    command_raw(cmd as u8);
-    read(&DATA_PORT)
-}
+pub static DATA_PORT: IOPort = unsafe { IOPort::new(0x60) };
+pub static STATUS_PORT: IOPort = unsafe { IOPort::new(0x64) };
+pub static COMMAND_PORT: IOPort = unsafe { IOPort::new(0x64) };
 
-/// Sends a raw controller command code
-fn command_raw(cmd: u8) {
-    write(&COMMAND_PORT, cmd);
-}
+pub const WAIT_TIMEOUT: u16 = 1000;
 
-/// Writes the given value to the data port
-pub fn write_data(value: u8) {
-    write(&DATA_PORT, value);
+bitflags! {
+    pub struct StatusFlags: u8 {
+        /// If the output buffer from the controller is full (data can be read)
+        const OUTPUT_FULL = 1 << 0;
+        /// If the input buffer to the controller is full (data cannot be written)
+        const INPUT_FULL = 1 << 1;
+        /// If the current output from the controller is from the second port
+        const OUTPUT_PORT_2 = 1 << 5;
+    }
 }
 
-/// Reads from the data port
-pub fn read_data() -> Option<u8> {
-    read(&DATA_PORT)
+/// Represents an error returned by PS/2
+#[derive(Copy, Clone, Debug)]
+pub enum Ps2Error {
+    NoData,
+    DeviceUnavailable,
 }
 
-/// Writes to the given port, or waits until available
-pub fn write(port: &IOPort, value: u8) {
-    wait_write();
-    port.write(value);
+/// Sends a controller command without a return
+pub fn command(cmd: commands::ControllerCommand) -> Result<(), Ps2Error> {
+    write(&COMMAND_PORT, cmd as u8)
```

Comment:
Maybe this could be put into the command module, and could be 'send', and the others 'send_ret', etc.